src/scheduler.ml

```diff
@@ -54,24 +122,52 @@ end = struct
     | Some x -> x
 
   let wait_unix () =
-    let pid, status = Unix.wait () in
-    (resolve_and_remove_job pid, status)
+    Unix.wait ()
 
   let wait =
     if Sys.win32 then
       wait_win32
     else
       wait_unix
+
+  let rec wait_to_chan chan =
+    try
+      let pid, status = wait () in
+      Event.(sync (send chan (pid, status)));
+      wait_to_chan chan
+    with
+    | Unix.Unix_error (err, _, _) as exn ->
+      if err = Unix.ECHILD then begin
+        Thread.delay 0.001;
```

Comment:
We should avoid busy polling. We can use a condition variable to detect when we should call `Unix.wait` and when we should wait for a process to be started.